### PR TITLE
add: support to set and reset genbank location types in a round trip 

### LINKED ISSUE #166: Support all of the sample files in the iGEM distribution
- Big, but not complicated
- 100s of GenBank conversions
- First step is to convert the GenBank files into SBOL and verify that the SBOL files are similar enough, although they don’t have to be identical
- Second step is to take the SBOL file that is produced, and convert that to a giant GenBank file and make sure that has substantially the same information as the current GenBank file that is produced

Note: This follows after #150, #164, #165

### PR DESCRIPTION
**Pull request by mohitdmak**
Closes #166 

 Relevant Issue created in SBOL-Specification repo: https://github.com/SynBioDex/SBOL-specification/issues/491
 The issue regarding the pySBOL3 bug on providing identity to Location class: https://github.com/SynBioDex/pySBOL3/issues/414
 Bug Fix for the above issue merged by @tcmitchell: https://github.com/SynBioDex/pySBOL3/pull/415

### Inappropriate GenBank Locus names for SBOL's display id:
* There are cases wherein GenBank locus names that are loosely created don't conform to SBOL's requirements for them being *display_id*s of SBOL3 components (i.e: strings being alphanumeric and not starting with a number). Thus, directly storing locus names as *display_id*s of components led to parsing errors begin thrown by SBOL3.
* To resolve this, we use SBOL's **string_to_display_id()** method which conveniently converts a string to conform to SBOL's requirements by either substituting non-alphanumeric characters with alphanumeric characters or removing them.
* However, we also store the original Locus name in a separate string extraneous property field in the extended GenBank component class, which is reset back during a round-trip conversion.
* A unit test for the same has been added along with the test-files

### Feature Location Endpoints Types:
* GenBank Feature Locations support the starting and ending points to be "fuzzy" or "exact"; i.e, the endpoint can be exact or denote that all points after that point are valid endpoints. Similarly, the start point can be exact or denote that all points before that point are valid start points.
* This type of location endpoint is not currently supported by SBOL3, as a workaround we store all such features having "fuzzy" features separately as extraneous property of a component, using the `Component_GenBank_Extension`'s **fuzzy_features** field, which is an *OwnedObject* ListProperty, which can store the list of all such fuzzy features.
* The reason behind not storing these features directly in SBOL3 SequenceFeatures is that in the absence of an SBOL3 field depicting the "exactness" or "fuzziness" of a location endpoint, the stored feature would be incompletely or inaccurately interpreted.
* With respect to this shortcoming, an issue has been created in the *SBOL-Specification Repo* to consider adding support to store fuzzy endpoints in Feature Locations [here](https://github.com/SynBioDex/SBOL-specification/issues/491).

### Round-tripping Fuzzy Features:
* Although conversion from GenBank to SBOL3 would "ignore" (store separately as extraneous property) features with fuzzy location endpoints, in a round trip, we need to preserve the features as the final GenBank output must be identical to the supplied file.
* For this purpose, the "Range" object has been extended as `Range_GenBank_Extension`, which stores *start_position* and *end_points* Integer fields, which are mapped to their corresponding GenBank Location Type operators (**BeforePosition**, **ExactPosition**, **AfterPosition**), using maps defined in the main converter class.
* `SBOL_LOCATION_POSITION` and `GENBANK_LOCATION_POSITION` map these integers to and from their GenBank operators. Also, a `build_range_extension` helper method is created to register the builder for this custom extension on the URI `sbol3.SBOL_RANGE`.
* The feature handling helper methods then keep track of GenBank features having fuzzy locations and store them in the extraneous "fuzzy_features" property instead of the corresponding SBOL3 component's features.
* In the round trip to GenBank, all features (every SBOL3 component's features, and all fuzzy_features) are considered to be converted to GenBank.

## TODO:

- [x]  Extend SBOL's **Location** class instead of **Range** class to store extraneous *fuzzy features*.
- [x]  Remove the [temporary workaround](https://github.com/mohitdmak/SBOL-utilities/pull/2) for the [pySBOL3 bug](https://github.com/SynBioDex/pySBOL3/issues/414) once a new release is made.
  - This task is not complete, and it cannot be completed at this time. It requires a new release of pySBOL3, followed by an update to SBOL-utilities to incorporate that new release. 
  - Relevant issue for the same is created at https://github.com/SynBioDex/SBOL-utilities/issues/187

### CODE REVIEW

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:178]: I'm thinking this should be a child of sbol3.Location instead of sbol3.Range, because something that treats an inexact range as an exact range will have errors.

**Review by jakebeal — CHANGES_REQUESTED** [review]

**Comment by tcmitchell** [comment]
I've looked at the code and it looks good to me. I haven't yet seen a change that addresses the change requested by @jakebeal so I am holding off on approval.

I thought the handing of fuzzy features was clear, and I thought the handling of the genbank name vs. SBOL display id was clear. Nice job coding those up!

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: With the new workaround merged for the pySBOL3 merge, this seems to have been appropriately implemented now.

**Comment by mohitdmak** [comment]
@tcmitchell @jakebeal With the workaround implemented (and noting it to be removed once a new pySBOL3 release is made), and the handling of improper display ids from GenBank in place, I think this PR is good to merge

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (tcmitchell) [reply sbol_utilities/sbol3_genbank_conversion.py]: @jakebeal I'm leaving it to you to resolve this requested change. I agree with @mohitdmak that the change we collaborated on yesterday resolves this issue.

**Review by tcmitchell — APPROVED** [review]

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:78]: - [ ] Please delete this commented code before merging.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:134]: - [ ] Please remove comment redundancy and delete commented code before merging.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:196]: - [ ] Please delete commented code before merging.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:219]: - [ ] Please add an issue for removing the workaround after the next pySBOL3 release.

**Review by jakebeal — APPROVED** [review]
I have left tasks about a few bits of comment to delete and an issue to file before you finish the merge. I don't think there's any need to re-review after you do those tasks.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Done

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Done

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Done

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py:187]: Done, created at https://github.com/SynBioDex/SBOL-utilities/issues/187

**Comment by mohitdmak** [comment]
@tcmitchell @jakebeal the redundant comments are removed, and I have also added a commit after running an autopep8 formatted on the whole converter file code as well. This should be good to merge.

**Review by jakebeal — APPROVED** [review]

**Pull request merged**
